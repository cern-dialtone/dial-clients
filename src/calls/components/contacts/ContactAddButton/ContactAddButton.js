import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { Icon } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import dialBackendApi from 'services/api';

import styles from './ContactAddButton.module.css';

function ContactAddButton({ contact }) {
  const [hasContact, setHasContact] = useState(false);
  const [isLoading, setIsLoading] = useState(true);

  const contacts = useSelector(state => state.contacts.getContacts.contacts);
  const added = useSelector(state => state.contacts.addContacts.added);
  const removed = useSelector(state => state.contacts.removeContacts.removed);
  const dispatch = useDispatch();
  const addContact = async () => {
    await dispatch(dialBackendApi().addUserContact(contact));
    setHasContact(true);
  };
  const removeContact = async () => {
    await dispatch(dialBackendApi().removeUserContact(contact.username));
    setHasContact(false);
  };

  useEffect(() => {
    const ids =
      contacts === undefined || contacts.contacts === undefined
        ? []
        : contacts.contacts.map(a => a.username);

    if (ids.includes(contact.username) || ids.includes(contact.username)) {
      setHasContact(true);
    } else {
      setHasContact(false);
    }
    setIsLoading(false);
  }, [added, removed, contacts, contact]);

  if (isLoading) {
    return <Icon loading name="sync" size="small" />;
  }

  if (hasContact) {
    return (
      <Icon
        data-testid="HasContactIcon"
        name="star"
        className={styles.ContactAddButton}
        size="big"
        color="yellow"
        onClick={removeContact}
      />
    );
  }

  return (
    <Icon
      data-testid="HasNoContactIcon"
      name="star"
      className={styles.ContactAddButton}
      size="big"
      color="grey"
      onClick={addContact}
    />
  );
}

ContactAddButton.propTypes = {
  contact: PropTypes.shape({
    personId: PropTypes.string.isRequired
  }).isRequired
};

export default ContactAddButton;
