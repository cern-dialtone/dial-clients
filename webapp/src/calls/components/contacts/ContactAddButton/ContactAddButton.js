import React, { Component } from "react";
import { Icon } from "semantic-ui-react";
import { logMessage } from "common/utils/logs";
import PropTypes from "prop-types";
import styles from "./ContactAddButton.module.css";

class ContactAddButton extends Component {
  static propTypes = {
    contact: PropTypes.object.isRequired,
    contacts: PropTypes.array.isRequired,
    addUserContact: PropTypes.func.isRequired,
    removeUserContact: PropTypes.func.isRequired,
    getUserContacts: PropTypes.func.isRequired
  };

  state = {
    loading: true,
    deleting: true,
    hasContact: false
  };

  async componentDidMount() {
    const { getUserContacts, contact } = this.props;
    this.setState({ loading: true });
    const newContacts = await getUserContacts();
    if (newContacts && newContacts.payload && newContacts.payload.result) {
      let ids = newContacts.payload.result.map(a => a.personId.toString());
      if (
        ids.includes(contact.personId) ||
        ids.includes(contact.personId.toString())
      ) {
        this.setState({ hasContact: true });
      }
    }
    this.setState({ loading: false });
  }

  addContactAction = async () => {
    const { addUserContact, contact, getUserContacts } = this.props;
    this.setState({ loading: true });
    await addUserContact(contact);
    this.setState({ loading: false });
    this.setState({ hasContact: true });
    getUserContacts();
  };

  removeContactAction = async () => {
    const { removeUserContact, contact, getUserContacts } = this.props;
    logMessage(`Removing contact...`);
    this.setState({ loading: true });
    await removeUserContact(contact.personId);
    this.setState({ loading: false });
    this.setState({ hasContact: false });
    getUserContacts();
  };

  render() {
    const { loading, hasContact } = this.state;

    if (loading) {
      return <Icon loading name="sync" size={"small"} />;
    }

    if (hasContact) {
      return (
        <Icon
          name="star"
          className={styles.ContactAddButton}
          size="big"
          color={"yellow"}
          onClick={this.removeContactAction}
        />
      );
    }

    return (
      <Icon
        name="star"
        className={styles.ContactAddButton}
        size="big"
        color={"grey"}
        onClick={this.addContactAction}
      />
    );
  }
}

export default ContactAddButton;
