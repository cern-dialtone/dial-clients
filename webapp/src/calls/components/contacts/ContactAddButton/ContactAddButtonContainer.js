import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { addUserContact, getUserContacts, removeUserContact } from "calls/actions/contacts";
import ContactAddButton from "calls/components/contacts/ContactAddButton/ContactAddButton";


function mapStateToProps({ calls }) {
  return {
    contacts: calls.contacts.getContacts.contacts,
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({
    addUserContact,
    removeUserContact,
    getUserContacts,
  }, dispatch);
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
  )(ContactAddButton);