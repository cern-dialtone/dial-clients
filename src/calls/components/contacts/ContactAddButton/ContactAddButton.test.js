import React from 'react';
import { render, waitForElement } from '@testing-library/react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { createHashHistory } from 'history';
import { routerMiddleware } from 'connected-react-router';
import ContactAddButton from './ContactAddButton';

const history = createHashHistory();
const middlewares = [thunk, routerMiddleware(history)];
const mockStore = configureMockStore(middlewares);

describe('ContactAddButton tests', () => {
  let store;
  let initialState;

  beforeEach(() => {
    initialState = {
      contacts: {
        getContacts: {
          contacts: []
        },
        addContacts: {
          added: false
        },
        removeContacts: {
          removed: false
        }
      }
    };
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  it('renders the component with not added contact', async () => {
    const contact = {
      personId: '222333'
    };
    const { asFragment, queryByTestId, getByTestId } = render(
      <Provider store={store}>
        <ContactAddButton contact={contact} />
      </Provider>
    );
    await waitForElement(() => getByTestId('HasNoContactIcon'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the component with added contact', async () => {
    const contact = {
      personId: '222333',
      username: 'lesknope'
    };
    initialState.contacts.getContacts.contacts.contacts = [contact];
    const { asFragment, queryByTestId, getByTestId } = render(
      <Provider store={store}>
        <ContactAddButton contact={contact} />
      </Provider>
    );
    await waitForElement(() => getByTestId('HasContactIcon'));
    expect(asFragment()).toMatchSnapshot();
  });
});
