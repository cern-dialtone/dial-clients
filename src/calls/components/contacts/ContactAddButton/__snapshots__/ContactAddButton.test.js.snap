// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactAddButton tests renders the component with added contact 1`] = `
<DocumentFragment>
  <i
    aria-hidden="true"
    class="yellow star big icon ContactAddButton"
    data-testid="HasContactIcon"
  />
</DocumentFragment>
`;

exports[`ContactAddButton tests renders the component with not added contact 1`] = `
<DocumentFragment>
  <i
    aria-hidden="true"
    class="grey star big icon ContactAddButton"
    data-testid="HasNoContactIcon"
  />
</DocumentFragment>
`;
